src/libs/export.c

```diff
@@ -1055,18 +1055,26 @@ static void _update_formats_combobox(dt_lib_export_t *d)
   gtk_widget_set_sensitive(d->format, !empty);
 }
 
+static void _destroy_child(GtkWidget *widget, gpointer data)
+{
+  (void)data;  // avoid unreferenced-parameter warning
+  gtk_widget_destroy(widget);
+}
+
 static void _on_storage_list_changed(gpointer instance, dt_lib_module_t *self)
 {
   dt_lib_export_t *d = self->data;
   dt_imageio_module_storage_t *storage = dt_imageio_get_storage();
   dt_bauhaus_combobox_clear(d->storage);
 
-  GList *children, *iter;
-
-  children = gtk_container_get_children(GTK_CONTAINER(d->storage_extra_container));
-  for(iter = children; iter != NULL; iter = g_list_next(iter))
+#if 1  //TODO: this code needs careful testing before removing the old code in the 'else' branch
+  gtk_container_foreach(GTK_CONTAINER(d->storage_extra_container), _destroy_child, NULL);
```

Comment:
Sorry, my suggestion wasn't complete, but you can directly do
`gtk_container_foreach(GTK_CONTAINER(d->storage_extra_container), (GtkCallback)gtk_widget_destroy, NULL);`
There's actually already a dt_ui_container_destroy_children that uses that approach, so you could call that instead.

Reply:
Different number of args in the casting, hence a real no-no.  Calling `dt_ui_container_destroy_children` is the right thing to do here, but that function will need fixing to avoid future surprises.

Reply:
I've seen this usage in several places so assumed it was supported, but you've convinced me otherwise, so an adapter function is needed indeed. For the special case where the adapter reduces a `f(widget*,void*) to f(widget*)` it should be possible to use a generic adapter, that takes the widget and a pointer to the (standard gtk) function. So you'd call
`gtk_container_foreach(container, gtk_widget_function_single_parameter, gtk_widget_destroy);`
and `gtk_widget_function_single_parameter `simply does
```
void gtk_widget_function_single_parameter(GtkWidget *widget, gpointer data)
{
  void (*widget_function) (GtkWidget *) = data;
  widget_function(widget);
}
```
That way you don't have to write adapters for each of gtk_widget_destroy/show/hide/map/realize/queue_draw etc etc. So I kindof expected Gtk to have provided this generic adapter, but I can't find it.

And anyway, I'm sure you won't like the idea because there'd be no type checking on the second (function) parameter to foreach... ;-)

Reply:
regexp foreach.*gtk_widget gives 4 instances of the gtk_widget_destroy pattern, 3 of which have been in the code for years before I introduced the last one a few months ago. Oops?

Reply:
That's the nasty part about code with undefined behavior -- it can work just fine for many years, then a change in the compiler or the system makes it blow up....

Definitely a trade-off to be decided about on the type-safety versus lots-of-similar-instances front.

And then there are all the cases where a function returns a newly-allocated object which the caller never frees.  I keep finding more of them....  C++ smart pointers would be so useful here.